Client/Frontend/Home/ActivityStreamTopSitesCell.swift

```diff
@@ -0,0 +1,392 @@
+        }
+    }
+
+}
+
+struct ASHorizontalScrollCellUX {
+    static let TopSiteCellIdentifier = "TopSiteCell"
+    static let TopSiteItemSize = CGSize(width: 100, height: 100)
+    static let BackgroundColor = UIColor.whiteColor()
+    static let PageControlRadius: CGFloat = 3
+    static let PageControlSize = CGSize(width: 30, height: 15)
+    static let PageControlOffset = -5
+}
+
+/*
+ The View that describes the topSite cell that appears in the tableView.
+ */
+class ASHorizontalScrollCell: UITableViewCell {
+
+    lazy private var collectionView: UICollectionView = {
+        let layout  = HorizontalFlowLayout()
+        layout.itemSize = ASHorizontalScrollCellUX.TopSiteItemSize
+        layout.heightShouldChange = { [weak self] num in
+            self?.heightChanged(num)
+        }
+        let collectionView = UICollectionView(frame: CGRectZero, collectionViewLayout: layout)
+        collectionView.registerClass(TopSiteCell.self, forCellWithReuseIdentifier: ASHorizontalScrollCellUX.TopSiteCellIdentifier)
+        collectionView.backgroundColor = ASHorizontalScrollCellUX.BackgroundColor
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.pagingEnabled = true
+        return collectionView
+    }()
+
+    lazy private var pageControl: FilledPageControl = {
+        let pageControl = FilledPageControl()
+        pageControl.tintColor = UIColor.grayColor()
+        pageControl.indicatorRadius = ASHorizontalScrollCellUX.PageControlRadius
+        pageControl.userInteractionEnabled = false
+        return pageControl
+    }()
+
+    weak var parentTableView: UITableView?
+
+    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        backgroundColor = ASHorizontalScrollCellUX.BackgroundColor
+        contentView.addSubview(collectionView)
+        contentView.addSubview(pageControl)
+
+        collectionView.snp_makeConstraints { make in
+            make.edges.equalTo(contentView).priorityLow()
+        }
+
+        pageControl.snp_makeConstraints { make in
+            make.size.equalTo(ASHorizontalScrollCellUX.PageControlSize)
+            make.top.equalTo(collectionView.snp_bottom).offset(ASHorizontalScrollCellUX.PageControlOffset)
+            make.centerX.equalTo(self.snp_centerX)
+        }
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        let layout = collectionView.collectionViewLayout as! HorizontalFlowLayout
+        pageControl.pageCount = layout.numberOfPages
+        pageControl.hidden = pageControl.pageCount <= 1
+    }
+
+    func heightChanged(newHeight: Int) {
+        collectionView.snp_updateConstraints { make in
+            make.edges.equalTo(contentView).priorityLow()
+            make.height.equalTo(newHeight).priorityMedium()
+        }
+        self.layoutSubviews()
+        parentTableView?.reloadData()
```

Comment:
> But short of hard coding in default heights for each different device type we are stuck in the same boat as the previous TopSites. 

How about we hard code the height for the Top Sites section? We could have set of various heights for each trait collection (compact/regular). It would also let us be exact with the measurements from the designs without having to auto-infer anything. If hard coding the heights for the section makes everything else align nicely and fall into place architecturally it might make sense and simplify the code greatly.